FV3/atmos_cubed_sphere/tools/coarse_grained_restart_files.F90

```diff
@@ -364,27 +398,191 @@ subroutine coarse_grain_fv_tracer_restart_data_on_model_levels(Atm, mass)
+     real, intent(in) :: coarse_phalf(is_coarse:ie_coarse,js_coarse:je_coarse,1:npz+1)
+     real, intent(in) :: coarse_phalf_on_fine(is:ie,js:je,1:npz+1)
+     real, intent(in), dimension(is:ie,js:je,1:npz) :: masked_mass_weights, masked_area_weights
+
+     real, allocatable :: remapped(:,:,:)  ! Will re-use this to save memory
+
+     allocate(remapped(is:ie,js:je,1:npz))
+
+     call remap_edges_along_x(Atm%u(is:ie,js:je+1,1:npz), &
+       phalf(is-1:ie+1,js-1:je+1,1:npz+1), &
+       Atm%gridstruct%dx(is:ie,js:je+1), &
+       Atm%ptop, &
+       Atm%coarse_graining%restart%u)
+     call remap_edges_along_y(Atm%v(is:ie+1,js:je,1:npz), &
+       phalf(is-1:ie+1,js-1:je+1,1:npz+1), &
+       Atm%gridstruct%dy(is:ie+1,js:je), &
+       Atm%ptop, &
+       Atm%coarse_graining%restart%v)
+
+     call vertically_remap_field(phalf(is:ie,js:je,1:npz+1), Atm%pt(is:ie,js:je,1:npz), coarse_phalf_on_fine, Atm%ptop, remapped)
+     call weighted_block_average(masked_mass_weights, remapped, Atm%coarse_graining%restart%pt)
+
+     if (.not. Atm%flagstruct%hydrostatic) then
+       call vertically_remap_field(phalf(is:ie,js:je,1:npz+1), Atm%w(is:ie,js:je,1:npz), coarse_phalf_on_fine, Atm%ptop, remapped)
+       call weighted_block_average(masked_mass_weights, remapped, Atm%coarse_graining%restart%w)
+       call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), Atm%delz(is:ie,js:je,1:npz), Atm%coarse_graining%restart%delz)
+       if (Atm%flagstruct%hybrid_z) then
+          call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), Atm%ze0(is:ie,js:je,1:npz), Atm%coarse_graining%restart%ze0)
+       endif
+     endif
+
+     call weighted_block_average(Atm%gridstruct%area(is:ie,js:je), Atm%phis(is:ie,js:je), Atm%coarse_graining%restart%phis)
+
+     if (Atm%coarse_graining%write_coarse_agrid_vel_rst) then
+       call vertically_remap_field(phalf(is:ie,js:je,1:npz+1), Atm%ua(is:ie,js:je,1:npz), coarse_phalf_on_fine, Atm%ptop, remapped)
+       call weighted_block_average(masked_mass_weights, remapped, Atm%coarse_graining%restart%ua)
+       call vertically_remap_field(phalf(is:ie,js:je,1:npz+1), Atm%va(is:ie,js:je,1:npz), coarse_phalf_on_fine, Atm%ptop, remapped)
+       call weighted_block_average(masked_mass_weights, remapped, Atm%coarse_graining%restart%va)
+     endif
+  end subroutine coarse_grain_fv_core_restart_data_on_pressure_levels
+
+  subroutine coarse_grain_fv_tracer_restart_data_on_pressure_levels( &
+     Atm, phalf, coarse_phalf_on_fine, masked_mass_weights, masked_area_weights)
+     type(fv_atmos_type), intent(inout) :: Atm
+     real, intent(in) :: phalf(is-1:ie+1,js-1:je+1,1:npz+1)
+     real, intent(in) :: coarse_phalf_on_fine(is:ie,js:je,1:npz+1)
+     real, intent(in), dimension(is:ie,js:je,1:npz) :: masked_mass_weights, masked_area_weights
+
+     real, allocatable :: remapped(:,:,:)
+     character(len=64) :: tracer_name
+     integer :: n_tracer
+ 
+     allocate(remapped(is:ie,js:je,1:npz))
```

Comment:
double checking... this will be automatically deallocated at end of subroutine right?

Reply:
Yes, I believe it should be.